homeassistant/components/vacuum/mqtt.py

```diff
@@ -0,0 +1,251 @@
+        self._payload_turn_off = "turn_off"
+        self._payload_return_to_base = "return_to_base"
+        self._payload_stop = "stop"
+        self._payload_clean_spot = "clean_spot"
+        self._payload_locate = "locate"
+        self._payload_start_pause = "start_pause"
+        self._retain = False
+
+        self._state_template = state_template
+
+        self._state = False
+        self._status = 'Unknown'
+        self._battery_level = 0
+
+    @asyncio.coroutine
+    def async_added_to_hass(self):
+        """Subscribe MQTT events.
+
+        This method is a coroutine.
+        """
+        @callback
+        def message_received(topic, payload, qos):
+            """Handle new MQTT message."""
+            try:
+                payload = json.loads(payload)
+            except ValueError:  # e.g. json.decoder.JSONDecodeError:
+                _LOGGER.warning("JSONDecodeError in vacuum.mqtt payload: %s",
+                                payload)
+                return
+
+            battery_level = payload.get(ATTR_BATTERY_LEVEL)
+            if battery_level is not None:
+                self._battery_level = int(battery_level)
+
+            charging = payload.get("charging")
+            state = payload.get(ATTR_STATE)
+            if payload[ATTR_STATE]:
+                if state == "cleaning":
+                    self._state = True
+                    self._status = "Cleaning"
+                if state == "docked":
+                    self._state = False
+                    if charging:
+                        self._status = "Docked & Charging"
+                    else:
+                        self._status = "Docked"
+                if state == "stopped":
+                    self._state = False
+                    self._status = "Stopped"
+
+            self.hass.async_add_job(self.async_update_ha_state())
+
+        yield from mqtt.async_subscribe(
+            self.hass, self._state_topic, message_received, self._qos)
+
+    @property
+    def name(self):
+        """Return the name of the vacuum."""
+        return self._name
+
+    @property
+    def icon(self):
+        """Return the icon for the vacuum."""
+        return DEFAULT_ICON
+
+    @property
+    def should_poll(self):
+        """No polling needed for an MQTT vacuum."""
+        return False
+
+    @property
+    def is_on(self):
+        """Return true if vacuum is on."""
+        return self._state
+
+    @property
+    def status(self):
+        """Return the status of the vacuum."""
+        if self.supported_features & SUPPORT_STATUS == 0:
+            return
+
+        return self._status
+
+    @property
+    def battery_level(self):
+        """Return the status of the vacuum."""
+        if self.supported_features & SUPPORT_BATTERY == 0:
+            return
+
+        return max(0, min(100, self._battery_level))
+
+    @property
+    def device_state_attributes(self):
+        """Return device state attributes."""
+        return {}
```

Comment:
No need to overwrite this property if you don't return a dict with contents.

Reply:
Ah cool. I'll have to do more reading about what sort of things typically go in `device_state_attributes`